autonomi/src/client/high_level/vault/mod.rs

```diff
@@ -78,21 +144,203 @@ impl Client {
         secret_key: &VaultSecretKey,
         content_type: VaultContentType,
     ) -> Result<AttoTokens, VaultError> {
-        let scratch_address = ScratchpadAddress::new(secret_key.public_key());
-        info!("Writing to vault at {scratch_address:?}");
-
-        match self
-            .scratchpad_update(secret_key, content_type, &data)
-            .await
-        {
-            Ok(()) => Ok(AttoTokens::zero()),
-            Err(ScratchpadError::CannotUpdateNewScratchpad) => {
-                let (price, _) = self
-                    .scratchpad_create(secret_key, content_type, &data, payment_option)
-                    .await?;
-                Ok(price)
+        info!("Writing {} bytes to vault ...", data.len());
+        let mut total_cost = AttoTokens::zero();
+        let main_secret_key = MainSecretKey::new(secret_key.clone());
+
+        // scratchpad_derivations ordered by the collection order
+        let (mut cur_free_graphentry_derivation, mut scratchpad_derivations) = self
+            .current_capacity(
+                &main_secret_key,
+                DerivationIndex::from_bytes(VAULT_HEAD_DERIVATION_INDEX),
+            )
+            .await?;
+
+        let content_chunks = chunk_bytes(data);
+
+        info!(
+            "Current capacity is {}, meanwhile requiring {}",
+            scratchpad_derivations.len(),
+            content_chunks.len()
+        );
+
+        // claim more capacity if short of.
+        // Note: as the Scratchpad is `created on use`, hence during the `claim stage`,
+        //       NUM_OF_SCRATCHPADS_PER_GRAPHENTRY to be claimed in one newly created GraphEntry.
+        while scratchpad_derivations.len() < content_chunks.len() {
+            let (new_free_graphentry_derivation, new_scratchpad_derivations, grap_cost) = self
+                .expand_capacity(
+                    &main_secret_key,
+                    &cur_free_graphentry_derivation,
+                    payment_option.clone(),
+                )
+                .await?;
+            cur_free_graphentry_derivation = new_free_graphentry_derivation;
+            scratchpad_derivations.extend(&new_scratchpad_derivations);
+            total_cost = AttoTokens::from_atto(total_cost.as_atto() + grap_cost.as_atto());
+        }
+
+        for (i, content) in content_chunks.into_iter().enumerate() {
+            let sp_secret_key = main_secret_key
+                .derive_key(&DerivationIndex::from_bytes(scratchpad_derivations[i].1));
```

Comment:
Again access with [i] is dangerous. 
Maybe we could zip it into the iteration above? 
`(i, content, dx)`

Reply:
there is a check_and_create above `while scratchpad_derivations.len() < contents.len() {`
hence shall be safe here to access with index.

anyway, will change to `zip` to reduce the usage of `index access` as much as possible

Reply:
If extra cloning can avoid manual checks vs compile-time checks then I'd go for extra cloning. The rust compiler does an excellent job of ignoring `clone` when it can in release mode, I wouldn't worry about performance here.